## Topic: The title of our data set is IMdB's Top 250 movies of all time

### source: [https://www.kaggle.com/datasets/shauryasrivastava01/imdb\-top\-250\-movies\-of\-all\-time\-19212025](https://www.kaggle.com/datasets/shauryasrivastava01/imdb-top-250-movies-of-all-time-19212025)

The purpose of this data is to see out of the thousands of movies released, what 250 have become the most loved of all time. It ranges from the 1920s all the way to 2025, and the goal of this data set is to help industry professionals, writers, directors, and fans to see learn about these movies. 


In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('movies.csv')
df

,title,year,rating,votes,runtime,certificate,genre,gross_total
0,The Shawshank Redemption,1994,9.3,"3,168,010",142,R,Drama,$29.33M
1,The Godfather,1972,9.2,"2,212,884",175,R,"Crime, Drama",$250.93M
2,The Dark Knight,2008,9.1,"3,145,646",152,PG-13,"Action, Crime, Drama",$1008.48M
3,The Godfather Part II,1974,9.0,"1,486,878",202,R,"Crime, Drama",$48.20M
4,12 Angry Men,1957,9.0,"977,359",96,Unrated,"Crime, Drama",$0.01M
...,...,...,...,...,...,...,...,...
245,Drishyam,2015,8.2,"107,875",163,Unrated,"Crime, Drama, Mystery",NaN
246,The Grapes of Wrath,1940,8.1,"108,041",129,Unrated,Drama,$0.01M
247,To Be or Not to Be,1942,8.1,"49,586",99,Unrated,"Comedy, Romance, War",$0.00M
248,Gangs of Wasseypur,2012,8.2,"113,517",321,Unrated,"Action, Comedy, Crime",$4.38M


#### Basic Data Exploration



In [5]:
df.head(5)

,title,year,rating,votes,runtime,certificate,genre,gross_total
0,The Shawshank Redemption,1994,9.3,"3,168,010",142,R,Drama,$29.33M
1,The Godfather,1972,9.2,"2,212,884",175,R,"Crime, Drama",$250.93M
2,The Dark Knight,2008,9.1,"3,145,646",152,PG-13,"Action, Crime, Drama",$1008.48M
3,The Godfather Part II,1974,9.0,"1,486,878",202,R,"Crime, Drama",$48.20M
4,12 Angry Men,1957,9.0,"977,359",96,Unrated,"Crime, Drama",$0.01M


In [6]:
df.tail(5)

,title,year,rating,votes,runtime,certificate,genre,gross_total
245,Drishyam,2015,8.2,"107,875",163,Unrated,"Crime, Drama, Mystery",NaN
246,The Grapes of Wrath,1940,8.1,"108,041",129,Unrated,Drama,$0.01M
247,To Be or Not to Be,1942,8.1,"49,586",99,Unrated,"Comedy, Romance, War",$0.00M
248,Gangs of Wasseypur,2012,8.2,"113,517",321,Unrated,"Action, Comedy, Crime",$4.38M
249,Höstsonaten,1978,8.1,"41,972",99,PG,"Drama, Music",$0.04M


In [7]:
df.shape

(250, 8)

In [8]:
df.columns

Index(['title', 'year', 'rating', 'votes', 'runtime', 'certificate', 'genre',
       'gross_total'],
      dtype='object')

#### Data Types and Structures



In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   title        250 non-null    object 
 1   year         250 non-null    int64  
 2   rating       250 non-null    float64
 3   votes        250 non-null    object 
 4   runtime      250 non-null    int64  
 5   certificate  250 non-null    object 
 6   genre        250 non-null    object 
 7   gross_total  246 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 15.8+ KB


In [10]:
df.select_dtypes(include='number')

,year,rating,runtime
0,1994,9.3,142
1,1972,9.2,175
2,2008,9.1,152
3,1974,9.0,202
4,1957,9.0,96
...,...,...,...
245,2015,8.2,163
246,1940,8.1,129
247,1942,8.1,99
248,2012,8.2,321


In [11]:
df.select_dtypes(include='object')

,title,votes,certificate,genre,gross_total
0,The Shawshank Redemption,"3,168,010",R,Drama,$29.33M
1,The Godfather,"2,212,884",R,"Crime, Drama",$250.93M
2,The Dark Knight,"3,145,646",PG-13,"Action, Crime, Drama",$1008.48M
3,The Godfather Part II,"1,486,878",R,"Crime, Drama",$48.20M
4,12 Angry Men,"977,359",Unrated,"Crime, Drama",$0.01M
...,...,...,...,...,...
245,Drishyam,"107,875",Unrated,"Crime, Drama, Mystery",NaN
246,The Grapes of Wrath,"108,041",Unrated,Drama,$0.01M
247,To Be or Not to Be,"49,586",Unrated,"Comedy, Romance, War",$0.00M
248,Gangs of Wasseypur,"113,517",Unrated,"Action, Comedy, Crime",$4.38M


I think votes and gross total should be converted to numerical because the former is a numerical count of votes on the website and gross total is how much money the movie made. They're both typically numerical values


#### Descriptive Summary



In [12]:
df.describe

<bound method NDFrame.describe of                         title  year  rating      votes  runtime certificate  \
0    The Shawshank Redemption  1994     9.3  3,168,010      142           R   
1               The Godfather  1972     9.2  2,212,884      175           R   
2             The Dark Knight  2008     9.1  3,145,646      152       PG-13   
3       The Godfather Part II  1974     9.0  1,486,878      202           R   
4                12 Angry Men  1957     9.0    977,359       96     Unrated   
..                        ...   ...     ...        ...      ...         ...   
245                  Drishyam  2015     8.2    107,875      163     Unrated   
246       The Grapes of Wrath  1940     8.1    108,041      129     Unrated   
247        To Be or Not to Be  1942     8.1     49,586       99     Unrated   
248        Gangs of Wasseypur  2012     8.2    113,517      321     Unrated   
249               Höstsonaten  1978     8.1     41,972       99          PG   

                 

In [13]:
df.describe()
### This statistics tell us that for the numerical columns the average year was 1988, raing 8.3, and runtime was 131 minutes

,year,rating,runtime
count,250.000000,250.000000,250.000000
mean,1988.972000,8.310400,131.932000
std,25.393197,0.231747,36.547589
min,1921.000000,8.000000,45.000000
25%,1973.000000,8.100000,108.250000
50%,1995.500000,8.200000,127.500000
75%,2009.000000,8.400000,147.750000
max,2025.000000,9.300000,374.000000


In [14]:
### No values jump out to me as unusual or inconsistent based off the summary

#### Missing or duplicate data

In [15]:
df.isnull()

,title,year,rating,votes,runtime,certificate,genre,gross_total
0,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...
245,False,False,False,False,False,False,False,True
246,False,False,False,False,False,False,False,False
247,False,False,False,False,False,False,False,False
248,False,False,False,False,False,False,False,False


#### The only missing is the gross_total for row 245

In [16]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
245    False
246    False
247    False
248    False
249    False
Length: 250, dtype: bool

#### The missing data might affect analysis by affecting how we evaluate the gross total's summary statistics and how we evaluate the movie in comparison to the others

6. Add new column using existing column

In [17]:
df['different_ratings']= df['rating'].astype(str) + df['certificate']
df['different_ratings'].head(5)

0          9.3R
1          9.2R
2      9.1PG-13
3          9.0R
4    9.0Unrated
Name: different_ratings, dtype: object

This column represents the movies ratings in terms how well received it was critically along with their rating in terms of subject matter. This gives us a better understanding of if subject matter and the rating associated with it impacts critical acclaim.



### 7. Filter the Data

In [18]:
df.corr(numeric_only=True)

,year,rating,runtime
year,1.000000,0.065769,0.152417
rating,0.065769,1.000000,0.269314
runtime,0.152417,0.269314,1.000000


In [19]:
df.isnull().sum()

title                0
year                 0
rating               0
votes                0
runtime              0
certificate          0
genre                0
gross_total          4
different_ratings    0
dtype: int64

In [20]:
df['gross_total']=df['gross_total'].fillna(df['gross_total'].mode()[0])
df.tail(5)

,title,year,rating,votes,runtime,certificate,genre,gross_total,different_ratings
245,Drishyam,2015,8.2,"107,875",163,Unrated,"Crime, Drama, Mystery",$0.01M,8.2Unrated
246,The Grapes of Wrath,1940,8.1,"108,041",129,Unrated,Drama,$0.01M,8.1Unrated
247,To Be or Not to Be,1942,8.1,"49,586",99,Unrated,"Comedy, Romance, War",$0.00M,8.1Unrated
248,Gangs of Wasseypur,2012,8.2,"113,517",321,Unrated,"Action, Comedy, Crime",$4.38M,8.2Unrated
249,Höstsonaten,1978,8.1,"41,972",99,PG,"Drama, Music",$0.04M,8.1PG


In [21]:
df.isnull().sum()

title                0
year                 0
rating               0
votes                0
runtime              0
certificate          0
genre                0
gross_total          0
different_ratings    0
dtype: int64

In [22]:
def yearby_decade(year): 
    if year>= 2020: 
        return '2020s'
    elif year>= 2010: 
        return '2010s' 
    elif year>= 2000: 
        return '2000s'
    elif year>= 1990:
        return '1990s'
    elif year>= 1980: 
        return '1980s'
    elif year>= 1970: 
        return '1970s'
    else: 
        return '1960s or before'

In [23]:
df['yearby_decade']=df['year'].apply(yearby_decade)
df

,title,year,rating,votes,runtime,certificate,genre,gross_total,different_ratings,yearby_decade
0,The Shawshank Redemption,1994,9.3,"3,168,010",142,R,Drama,$29.33M,9.3R,1990s
1,The Godfather,1972,9.2,"2,212,884",175,R,"Crime, Drama",$250.93M,9.2R,1970s
2,The Dark Knight,2008,9.1,"3,145,646",152,PG-13,"Action, Crime, Drama",$1008.48M,9.1PG-13,2000s
3,The Godfather Part II,1974,9.0,"1,486,878",202,R,"Crime, Drama",$48.20M,9.0R,1970s
4,12 Angry Men,1957,9.0,"977,359",96,Unrated,"Crime, Drama",$0.01M,9.0Unrated,1960s or before
...,...,...,...,...,...,...,...,...,...,...
245,Drishyam,2015,8.2,"107,875",163,Unrated,"Crime, Drama, Mystery",$0.01M,8.2Unrated,2010s
246,The Grapes of Wrath,1940,8.1,"108,041",129,Unrated,Drama,$0.01M,8.1Unrated,1960s or before
247,To Be or Not to Be,1942,8.1,"49,586",99,Unrated,"Comedy, Romance, War",$0.00M,8.1Unrated,1960s or before
248,Gangs of Wasseypur,2012,8.2,"113,517",321,Unrated,"Action, Comedy, Crime",$4.38M,8.2Unrated,2010s


### 8. Unique Values and Categories

In [24]:
df.select_dtypes(include='object')

,title,votes,certificate,genre,gross_total,different_ratings,yearby_decade
0,The Shawshank Redemption,"3,168,010",R,Drama,$29.33M,9.3R,1990s
1,The Godfather,"2,212,884",R,"Crime, Drama",$250.93M,9.2R,1970s
2,The Dark Knight,"3,145,646",PG-13,"Action, Crime, Drama",$1008.48M,9.1PG-13,2000s
3,The Godfather Part II,"1,486,878",R,"Crime, Drama",$48.20M,9.0R,1970s
4,12 Angry Men,"977,359",Unrated,"Crime, Drama",$0.01M,9.0Unrated,1960s or before
...,...,...,...,...,...,...,...
245,Drishyam,"107,875",Unrated,"Crime, Drama, Mystery",$0.01M,8.2Unrated,2010s
246,The Grapes of Wrath,"108,041",Unrated,Drama,$0.01M,8.1Unrated,1960s or before
247,To Be or Not to Be,"49,586",Unrated,"Comedy, Romance, War",$0.00M,8.1Unrated,1960s or before
248,Gangs of Wasseypur,"113,517",Unrated,"Action, Comedy, Crime",$4.38M,8.2Unrated,2010s


In [25]:
print(f'The unique values in the title column are:\n')
print(df['title'].unique())

The unique values in the title column are:

['The Shawshank Redemption' 'The Godfather' 'The Dark Knight'
 'The Godfather Part II' '12 Angry Men'
 'The Lord of the Rings: The Return of the King' 'Schindler&apos;s List'
 'The Lord of the Rings: The Fellowship of the Ring' 'Pulp Fiction'
 'Il buono, il brutto, il cattivo' 'The Lord of the Rings: The Two Towers'
 'Forrest Gump' 'Fight Club' 'Inception'
 'Star Wars: Episode V - The Empire Strikes Back' 'The Matrix'
 'GoodFellas' 'Interstellar' 'One Flew Over the Cuckoo&apos;s Nest'
 'Se7en' 'It&apos;s a Wonderful Life' 'The Silence of the Lambs'
 'Shichinin no samurai' 'Saving Private Ryan' 'The Green Mile'
 'Cidade de Deus' 'La vita è bella' 'Terminator 2: Judgment Day'
 'Star Wars' 'Back to the Future' 'Sen to Chihiro no kamikakushi'
 'The Pianist' 'Gladiator' 'Gisaengchung' 'Hotaru no haka' 'Psycho'
 'The Lion King' 'Seppuku' 'The Departed' 'Whiplash' 'The Prestige'
 'Kill Bill: The Whole Bloody Affair' 'American History X' 'Léon'
 'Spi

In [26]:
print(f'The unique values in the certficate column are:\n')
print(df['certificate'].unique())

The unique values in the certficate column are:

['R' 'PG-13' 'Unrated' 'PG' 'G' 'NC-17']


In [27]:
print(f'The unique values in the votes column are:\n')
print(df['votes'].unique())

The unique values in the votes column are:

['3,168,010' '2,212,884' '3,145,646' '1,486,878' '977,359' '2,152,126'
 '1,577,558' '2,193,491' '2,421,067' '887,227' '1,945,026' '2,479,948'
 '2,585,212' '2,796,318' '1,496,595' '2,233,701' '1,382,099' '2,496,091'
 '1,152,486' '2,006,146' '552,664' '1,711,212' '400,179' '1,628,543'
 '1,551,508' '866,611' '808,265' '1,277,496' '1,561,809' '1,442,692'
 '959,384' '1,006,347' '1,837,102' '1,150,711' '380,478' '781,795'
 '1,251,328' '89,994' '1,545,064' '1,143,440' '1,576,959' '30,141'
 '1,267,862' '1,344,671' '513,688' '315,853' '653,074' '1,020,643'
 '1,232,633' '1,058,161' '1,870,222' '284,271' '566,914' '382,606'
 '215,684' '766,922' '1,311,927' '1,427,780' '722,936' '1,105,828'
 '1,359,383' '441,636' '780,193' '260,009' '160,732' '237,768' '1,218,463'
 '258,580' '1,769,293' '168,898' '837,039' '70,730' '1,206,463'
 '1,432,148' '1,982,116' '699,365' '465,769' '1,166,702' '387,323'
 '286,903' '713,159' '1,156,620' '480,469' '549,702' '478,860'

In [28]:
print(f'The unique values in the genre column are:\n')
print(df['genre'].unique())

The unique values in the genre column are:

['Drama' 'Crime, Drama' 'Action, Crime, Drama' 'Adventure, Drama, Fantasy'
 'Biography, Drama, History' 'Adventure, Drama, Western' 'Drama, Romance'
 'Crime, Drama, Thriller' 'Action, Adventure, Sci-Fi'
 'Action, Adventure, Fantasy' 'Action, Sci-Fi' 'Biography, Crime, Drama'
 'Adventure, Drama, Sci-Fi' 'Crime, Drama, Mystery'
 'Drama, Family, Fantasy' 'Crime, Drama, Horror' 'Action, Drama'
 'Drama, War' 'Crime, Drama, Fantasy' 'Comedy, Drama, Romance'
 'Adventure, Comedy, Sci-Fi' 'Animation, Adventure, Family'
 'Biography, Drama, Music' 'Action, Adventure, Drama' 'Drama, Thriller'
 'Animation, Drama, War' 'Drama, Horror, Mystery'
 'Animation, Adventure, Drama' 'Drama, Mystery' 'Drama, Music'
 'Drama, Mystery, Sci-Fi' 'Action, Crime, Thriller'
 'Animation, Action, Adventure' 'Drama, Romance, War' 'Comedy, Drama'
 'Horror, Sci-Fi' 'Drama, Western' 'Drama, Mystery, Thriller'
 'Drama, Mystery, War' 'Action, Adventure' 'Drama, Film-Noir'
 'Drama, 

In [29]:
print(f'The unique values in the gross_total column are:\n')
print(df['gross_total'].unique())

The unique values in the gross_total column are:

['$29.33M' '$250.93M' '$1008.48M' '$48.20M' '$0.01M' '$1148.97M'
 '$322.16M' '$897.00M' '$213.93M' '$25.27M' '$944.74M' '$678.23M'
 '$101.32M' '$839.79M' '$550.02M' '$473.33M' '$47.08M' '$773.86M'
 '$109.13M' '$329.81M' '$11.50M' '$272.74M' '$1.09M' '$482.35M' '$286.80M'
 '$30.69M' '$230.10M' '$517.78M' '$775.40M' '$392.98M' '$360.18M'
 '$120.10M' '$465.52M' '$262.55M' '$4.70M' '$32.27M' '$979.16M' '$0.02M'
 '$292.10M' '$50.43M' '$109.68M' '$6.72M' '$23.88M' '$20.33M' '$690.82M'
 '$13.50M' '$4.74M' '$426.59M' '$23.34M' '$109.51M' '$426.08M' '$0.47M'
 '$37.92M' '$5.44M' '$0.06M' '$105.27M' '$527.40M' '$40.06M' '$714.84M'
 '$389.93M' '$2052.42M' '$77.67M' '$394.88M' '$0.34M' '$50.29M' '$0.97M'
 '$321.46M' '$0.14M' '$131.06M' '$225.93M' '$2799.44M' '$1085.43M'
 '$823.29M' '$52.27M' '$401.16M' '$405.32M' '$11.49M' '$17.62M' '$213.22M'
 '$33.81M' '$9.67M' '$60.26M' '$356.30M' '$64.42M' '$2.11M' '$1078.96M'
 '$5.48M' '$20.93M' '$482.47M' '$7.

In [30]:
print(f'The unique values in the different_rating column are:\n')
print(df['different_ratings'].unique())

The unique values in the different_rating column are:

['9.3R' '9.2R' '9.1PG-13' '9.0R' '9.0Unrated' '9.0PG-13' '8.9PG-13' '8.8R'
 '8.8PG-13' '8.7PG' '8.7R' '8.7PG-13' '8.6R' '8.6PG' '8.6Unrated'
 '8.6PG-13' '8.5PG' '8.5R' '8.5Unrated' '8.5G' '8.5PG-13' '8.8Unrated'
 '8.4R' '8.4G' '8.4PG-13' '8.4PG' '8.4Unrated' '8.7Unrated' '8.3G'
 '8.3Unrated' '8.3R' '8.3PG-13' '8.3PG' '8.3NC-17' '8.2R' '8.2PG'
 '8.2Unrated' '8.2PG-13' '8.2G' '8.1PG' '8.1R' '8.1PG-13' '8.1Unrated'
 '8.1G' '8.0PG']


In [31]:
print(f'The unique values in the star_rating column are:\n')
print(df['yearby_decade'].unique())

The unique values in the star_rating column are:

['1990s' '1970s' '2000s' '1960s or before' '2010s' '1980s' '2020s']


### 9. Grouping and Aggregation

In [32]:
grouped = df.groupby('certificate')
print("Type of grouped rating:", type(grouped))

Type of grouped rating: <class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [33]:
print("Average rating by runtime:")
print(df.groupby('certificate')['runtime'].mean())

Average rating by runtime:
certificate
G          118.277778
NC-17      102.000000
PG         119.473684
PG-13      144.294118
R          136.196078
Unrated    130.070175
Name: runtime, dtype: float64


In [34]:
### From the grouped data I learned that pg-13 and r films that are the top 250 of all time usually have longer run times stretching well past 2 hours. While NC-17 had the shortest run time on average

### 10. Summary and Conclusion

In this data set, we found that some of the movies included made less than a million dollars, yet due to their votes and high rating were still included in this list. We found this very interesting. Due to the genre column including technically having multiple values it made trying to print the unique values cluttered and somewhat difficult. A pattern we saw was, the higher the votes, the higher the gross total is for the most part.

In [35]:
df_cleaned = df.copy()
df_cleaned

,title,year,rating,votes,runtime,certificate,genre,gross_total,different_ratings,yearby_decade
0,The Shawshank Redemption,1994,9.3,"3,168,010",142,R,Drama,$29.33M,9.3R,1990s
1,The Godfather,1972,9.2,"2,212,884",175,R,"Crime, Drama",$250.93M,9.2R,1970s
2,The Dark Knight,2008,9.1,"3,145,646",152,PG-13,"Action, Crime, Drama",$1008.48M,9.1PG-13,2000s
3,The Godfather Part II,1974,9.0,"1,486,878",202,R,"Crime, Drama",$48.20M,9.0R,1970s
4,12 Angry Men,1957,9.0,"977,359",96,Unrated,"Crime, Drama",$0.01M,9.0Unrated,1960s or before
...,...,...,...,...,...,...,...,...,...,...
245,Drishyam,2015,8.2,"107,875",163,Unrated,"Crime, Drama, Mystery",$0.01M,8.2Unrated,2010s
246,The Grapes of Wrath,1940,8.1,"108,041",129,Unrated,Drama,$0.01M,8.1Unrated,1960s or before
247,To Be or Not to Be,1942,8.1,"49,586",99,Unrated,"Comedy, Romance, War",$0.00M,8.1Unrated,1960s or before
248,Gangs of Wasseypur,2012,8.2,"113,517",321,Unrated,"Action, Comedy, Crime",$4.38M,8.2Unrated,2010s


In [36]:
df_cleaned.to_csv('movies_filtered.csv', index=False)